src/core/qgsofflineediting.cpp

```diff
@@ -164,157 +165,189 @@ void QgsOfflineEditing::synchronize()
     if ( remoteName.endsWith( remoteNameSuffix ) )
       remoteName.chop( remoteNameSuffix.size() );
     const QgsVectorLayer::LayerOptions options { QgsProject::instance()->transformContext() };
 
-    std::unique_ptr<QgsVectorLayer> remoteLayer = std::make_unique<QgsVectorLayer>( remoteSource, remoteName, remoteProvider, options );
-    if ( remoteLayer->isValid() )
+    std::shared_ptr<QgsVectorLayer> remoteLayer = std::make_shared<QgsVectorLayer>( remoteSource, remoteName, remoteProvider, options );
+
+    if ( ! remoteLayer->isValid() )
+    {
+      QgsDebugMsgLevel( QStringLiteral( "Skipping offline layer %1 because it failed to recreate its corresponding remote layer" ).arg( offlineLayer->id() ), 4 );
+      continue;
+    }
+
+    // Rebuild WFS cache to get feature id<->GML fid mapping
+    if ( remoteLayer->providerType().contains( QLatin1String( "WFS" ), Qt::CaseInsensitive ) )
     {
-      // Rebuild WFS cache to get feature id<->GML fid mapping
-      if ( remoteLayer->providerType().contains( QLatin1String( "WFS" ), Qt::CaseInsensitive ) )
+      QgsFeatureIterator fit = remoteLayer->getFeatures();
+      QgsFeature f;
+      while ( fit.nextFeature( f ) )
       {
-        QgsFeatureIterator fit = remoteLayer->getFeatures();
-        QgsFeature f;
-        while ( fit.nextFeature( f ) )
-        {
-        }
       }
-      // TODO: only add remote layer if there are log entries?
+    }
+
+    // TODO: only add remote layer if there are log entries?
+    // apply layer edit log
+    const QString sql = QStringLiteral( "SELECT \"id\" FROM 'log_layer_ids' WHERE \"qgis_id\" = '%1'" ).arg( offlineLayer->id() );
+    const int layerId = sqlQueryInt( database.get(), sql, -1 );
+
+    if ( layerId == -1 )
+    {
+      QgsDebugMsgLevel( QStringLiteral( "Skipping offline layer %1 because it failed to determine the offline editing layer id" ).arg( offlineLayer->id() ), 4 );
+      continue;
+    }
 
-      QgsVectorLayer *offlineLayer = qobject_cast<QgsVectorLayer *>( layer );
-      if ( offlineLayer->isValid() )
+    remoteLayersByOfflineId.insert( layerId, remoteLayer );
+    offlineLayersByOfflineId.insert( layerId, offlineLayer );
+  }
+
+  QgsDebugMsgLevel( QStringLiteral( "Found %1 offline layers in total" ).arg( offlineLayersByOfflineId.count() ), 4 );
+
+  QMap<QPair<QString, QString>, std::shared_ptr<QgsTransactionGroup>> transactionGroups;
+  if ( useTransaction )
+  {
+    for ( std::shared_ptr<QgsVectorLayer> &remoteLayer : remoteLayersByOfflineId )
+    {
+      const QString connectionString = QgsTransaction::connectionString( remoteLayer->source() );
+      const QPair pair( remoteLayer->providerType(), connectionString );
+      std::shared_ptr<QgsTransactionGroup> transactionGroup = transactionGroups.value( pair );
+
+      if ( !transactionGroup.get() )
+        transactionGroup = std::make_shared<QgsTransactionGroup>( this );
```

Comment:
The transaction group should either have `this` as parent and be owned through qt parent/child relationship _or_ be a `shared_ptr` otherwise we walk on thin ice regarding double free (=potential crash).

As far as I can see the TG is only used through the lifetime of this function, so no parent sounds safe.

Reply:
thanks for the remark, I will remove the parent